servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ProxyConnectConnectionFactoryFilter.java

```diff
@@ -74,51 +68,63 @@ private ProxyFilter(final ConnectionFactory<ResolvedAddress, C> delegate) {
 
         @Override
         public Single<C> newConnection(final ResolvedAddress resolvedAddress) {
-            return delegate().newConnection(resolvedAddress).flatMap(c ->
-                // We currently only have access to a StreamingHttpRequester, which means we are forced to provide an
-                // HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the connection
-                // filters we use the default strategy which should offload everything to be safe.
-                c.request(defaultStrategy(),
-                            reqRespFactory.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
-                 .flatMap(response -> {
-                if (SUCCESSFUL_2XX.contains(response.status())) {
-                    final Channel channel = ((NettyConnectionContext) c.connectionContext()).nettyChannel();
-                    final SingleSource.Processor<C, C> processor = newSingleProcessor();
+            return delegate().newConnection(resolvedAddress).flatMap(c -> {
+                try {
+                    // We currently only have access to a StreamingHttpRequester, which means we are forced to provide
+                    // an HttpExecutionStrategy. Because we can't be sure if there is any blocking code in the
+                    // connection filters we use the default strategy which should offload everything to be safe.
+                    return c.request(defaultStrategy(), c.connect(connectAddress).addHeader(CONTENT_LENGTH, ZERO))
+                            .flatMap(response -> handleConnectResponse(c, response))
+                            // Close recently created connection in case of any error while it connects to the proxy
+                            // or cancellation:
+                            .recoverWith(t -> c.closeAsync().concat(failed(t)))
+                            .whenCancel(() -> c.closeAsync().subscribe());
```

Comment:
> we really do not know if the cancel has come due to the operation was canceled by the user or due to some operators sending a cancel for the previous source when they move on to the next source (eg: `concat()`).

Can any operator cancel after success? IIUC they cancel the previous source only for non-success/non-complete cases.

LMK if I need to revert `whenFinally` here to prevent closure on cancel after success.

Reply:
> Can any operator cancel after success? 

Yes they do, consider this `connFactory.newConnection().concat(executor.timer(1, MILLISECONDS)` (eg: to add a delay to respond to connect)

`concat()` uses `SequentialCancellable` which [cancels the old](https://github.com/apple/servicetalk/blob/master/servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/SequentialCancellable.java#L59-L62) `Cancellable` when the new `Cancellable` is received, which in this case will be after the successful completion of `connFactory.newConnection()`.

More generally, we should not assume anywhere that `cancel` is only received before `success()` as `Cancellable` and `Subscriber` code paths are concurrent.

Reply:
`whenCancel` will unconditionally execute the callback when `cancel` is called regardless if the connection has been delivered downstream. If we have already delivered the `connection` we shouldn't later close it (regardless if someone cancels or not). In addition to this being the expected control flow, the RS spec has some rules which discuss cancel being a no-op after a terminal signal is delivered [1][2].

`afterFinally(SingleTerminalSignalConsumer<T> doFinally)` happens to enforce "only a single callback will be executed" but may still result in invoking the `onCancel()` call back and also calling the downstream `Subscriber#onSuccess(...)` for the following reasons:
- Subscription can be invoked on a different thread
- Data/terminal signals may still be delivered after `cancel` [3]

So `afterFinally` is an improvement over `afterCancel`, but still isn't ideal because we may deliver a closed object (and/or invoke `closeAsync()` concurrently).

[1] https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#1.6
> If a Publisher signals either onError or onComplete on a Subscriber, that Subscriber’s Subscription MUST be considered cancelled.

[2] https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#3.7
> After the Subscription is cancelled, additional Subscription.cancel() MUST be NOPs.

[3] https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#2.8
> A Subscriber MUST be prepared to receive one or more onNext signals after having called Subscription.cancel() if there are still requested elements pending [see 3.12]. Subscription.cancel() does not guarantee to perform the underlying cleaning operations immediately.

Reply:
>>  TBH don't see where it cancels the old Cancellable.

Aah, you are correct. I misread under an older assumption that we `cancel()` the previous `Cancellable`. 

Anyways, for other reasons me and Scott mention, unconditional `close()` upon `cancel()` isn't correct.

Reply:
Updated to use `whenFinally` https://github.com/apple/servicetalk/pull/1002/commits/d2bf22f6b3dd9d9f8a82a7193fd159b4ca069d50

Reply:
> So afterFinally is an improvement over afterCancel, but still isn't ideal because we may deliver a closed object (and/or invoke closeAsync() concurrently).

Ok ya this seems to be a problem. Can we remove the close-on-cancel part for now? 

Connection lifetime is anyways a problem in such situation out of the context of this filter as mentioned in https://github.com/apple/servicetalk/pull/1002#discussion_r404474939. 

Lets fix the obvious issue of leaking connection for non-200 responses and then handle lifecycle on cancel/early termination later.

Reply:
Removed in https://github.com/apple/servicetalk/pull/1002/commits/57e470b5307862d543c41e8274b3fc5ee634cfa7 and created #1010.